taskflows/taskflows/container_pull_taskflow.py

```diff
@@ -0,0 +1,279 @@
+        countdown=countdown(cluster),
+        link=postprocess_job.s(user, cluster, image, job, folder, container))
+
+
+@cumulus.taskflow.task
+def postprocess_job(task, _, user, cluster, image, job, folder, container):
+    task.taskflow.logger.info('Finished pulling the container')
+
+    task.taskflow.logger.info('Processing the results of the pull.')
+    client = create_girder_client(
+        task.taskflow.girder_api_url, task.taskflow.girder_token)
+
+    # Refresh state of job
+    job = client.get('jobs/%s' % job['_id'])
+
+    upload_job_output_to_folder(cluster, job, girder_token=task.taskflow.girder_token)
+
+    output_items = list(client.listItem(folder['_id']))
+    output_file = None
+    for item in output_items:
+        if item['name'] == OUTPUT_FILE:
+            files = list(client.listFile(item['_id']))
+            if len(files) != 1:
+                log_and_raise(task, 'Expecting a single file under item, found: %s' % len(files))
+            output_file = files[0]
+            break
+
+    if output_file is None:
+        log_and_raise(task, 'Could not locate output file: %s' % OUTPUT_FILE)
+
+    # Download the output file
+    output_io = io.BytesIO()
+    client.downloadFile(output_file['_id'], output_io)
+
+    # Remove the folder
+    client.delete('folder/%s' % folder['_id'])
+
+    pull_json = json.loads(output_io.getvalue().decode('utf-8'))
+    image_uri = pull_json.get('imageUri')
+    # Convert size to GB
+    size = round(pull_json.get('size', 0) / 1.e9, 2)
+
+    _ensure_image_uri_is_valid(task, container, image_uri)
+
+    _post_image_to_database(client, container, image, image_uri, size)
+
+    task.taskflow.logger.info('Success!')
+
+
+def _ensure_image_uri_is_valid(task, container, image_uri):
+    if not image_uri:
+        log_and_raise(task, 'Image uri is empty')
+
+    # Raises an exception if the uri is not valid
+    if container == 'singularity':
+        if '.sif' not in image_uri:
+            log_and_raise(task, 'Invalid image uri: ' + str(image_uri))
+    else:
+        if ':' not in image_uri:
+            log_and_raise(task, 'Invalid image uri: ' + str(image_uri))
+
+
+def _extract_digest(container, image_uri):
+    if container == 'singularity':
+        # Get the hashsum from the file path
+        return os.path.basename(image_uri).replace('.sif', '')
+    else:
+        return image_uri.split(':')[1]
+
+
+def _post_image_to_database(client, container, image, image_uri, size):
+    repository = image.get('repository')
+    tag = image.get('tag')
+    digest = _extract_digest(container, image_uri)
+
+    body = {
+        'type': container,
+        'repository': repository,
+        'tag': tag,
+        'digest': digest,
+        'size': size
+    }
+
+    try:
+        client.post('images', body)
+    except HttpError as e:
+        # Just ignore the error if the image already exists
+        error_str = 'Image already exists'
+        if error_str not in e.responseText:
```

Comment:
This seems a little fragile to me. Why not use a status code, for example 409.

Reply:
This is now fixed.